File: support/build_functions.py
#!/bin/python
#
# builds marklogic builtins from api docs.
# point it at $dir/pubs/apidocs after unzipping the api docs

import os
import re
import json
import fnmatch
import xml.etree.ElementTree as ET

def make_increment_params(start=1):
	count = [start - 1]
	def create_increment_params(match):
		count[0] += 1
		return "${%d:\\$" % count[0]
	return create_increment_params

dir = "/path/to/MarkLogic_pubs/pubs/apidocs"

def break_apart_optional_params(sig):
	m = re.search(r"([^(]+)\((.+)?\)", sig)
	if m:
		function_name = m.group(1)
		if (m.group(2)):
			params = re.split(r",\s*", m.group(2))
		else:
			params = []

	required_params = []
	optional_params = []
	for param in params:
		if re.match(r"\[([^]]+)\],?", param):
			optional_params.append(re.sub(r"\[|\]", "", param))
		else:
			required_params.append(param)

	functions = []
	previous_optional_params = []

	# add the function with any required params
	functions.append("%s(%s)" % (function_name, ', '.join(required_params) ))

	for param in required_params:
		previous_optional_params.append(param)

	# add 1 function for each additional optional parameter
	for param in optional_params:
		previous_optional_params.append(param)
		functions.append("%s(%s)" % (function_name, ', '.join(previous_optional_params)))
	return functions

def get_sigs(filename, final_functions):
	root = ET.parse(filename)
	for e in root.findall('.//xhtml:table', namespaces={'xhtml':"http://www.w3.org/1999/xhtml"}):
		for se in e.findall('./xhtml:tr/xhtml:td/xhtml:a[@id]', namespaces={'xhtml':"http://www.w3.org/1999/xhtml"}):
			# sanitize the html into a function string
			sig = u''.join(e.itertext()).encode('utf-8').strip()
			sig = ''.join([i if ord(i) < 128 else ' ' for i in sig])
			sig = re.sub(r"\n", "", sig, re.DOTALL | re.M | re.U)
			sig = re.sub(r"\s+", " ", sig, re.DOTALL | re.M | re.U)
			sig = re.sub(r"\(\s+", "(", sig)
			sig = re.sub(r"\s+\)", ")", sig)
			sig = re.sub(r"\) as.*$", ")", sig)

			# if there are optional params, break this thing up into multiple instances
			signatures = break_apart_optional_params(sig)
			for signature in signatures:
				desc = re.sub(r" as\s.*?(?=,|\)$)+", "", signature)
				desc = re.sub(r"^[^(]+", "", desc)
				desc = re.sub(r",\s+", ",", desc)
				if (desc == "()"):
					desc = None

				# wrap the parameters with ${n:} where n is the parameter #
				if re.search(r"\$", signature):
					signature = re.sub(r"\$", make_increment_params(), signature)
					signature = re.sub(r",", "},", signature)
					signature = re.sub(r"\)$", "})", signature)

				function_name = re.match(r"([^(]+)\(", signature).group(1)

				obj = {
					"content": signature,
					"trigger": function_name
				}
				if (desc):
					obj['description'] = desc

				final_functions[signature] = obj
	return final_functions

functions = {}

exclude_list = ["All.html", "REST-Resources.html", "management-rest-api.html", "packageREST.html"]
for root, dirnames, filenames in os.walk(dir):
	for filename in fnmatch.filter(filenames, '*.html'):
		if (filename not in exclude_list and re.match(r"RESTclient", filename) == None):
			get_sigs(dir + "/" + filename, functions)

pruned_functions = []
for k in sorted(functions):
	pruned_functions.append(functions[k])

with open('../marklogic_builtins/ml-functions.json', 'w') as f:
	f.write(json.dumps(pruned_functions, sort_keys=True))

File: support/build_functions.rb
#!/bin/ruby

require 'typhoeus'
require 'nokogiri'
require 'json'

# HTTP_ROOT = "http://docs-ea.marklogic.com"
HTTP_ROOT = "http://pubs.marklogic.com:8011"

def break_apart_optional_params(sig, is_js)
	function_name = nil
	params = nil
	sig.gsub /([^(]+)\((.+)?\)/ do |m|
		function_name = $1
		if $2
			params = $2.split(/,\s*/)
		else
			params = [] unless $2
		end
	end

	required_params = []
	optional_params = []
	params.each do |param|

		param.gsub! /-([a-z])/ do |m|
			x = $1
			x.upcase
		end if is_js

		if param.match(/\[([^\]]+)\],?/)
			optional_params.push(param.gsub(/\[|\]/, ""))
		else
			required_params.push(param)
		end
	end

	functions = []
	previous_optional_params = []

	# add the function with any required params
	functions.push("#{function_name}(#{required_params.join(', ')})")

	required_params.each do |param|
		previous_optional_params.push(param)
	end

	# add 1 function for each additional optional parameter
	optional_params.each do |param|
		previous_optional_params.push(param)
		functions.push("#{function_name}(#{previous_optional_params.join(', ')})")
	end
	return functions
end

def get_function(hydra, url, final_functions, is_js = false)
	r = Typhoeus::Request.new("#{HTTP_ROOT}#{url}", method: :get)

	r.on_complete do |response|
		if response.code.to_i == 200
			desc = nil
			html = Nokogiri::XML::Document.parse(response.body)
			html.remove_namespaces!
			html.xpath('//div[@class="pjax_enabled"]/pre[1]').each do |e|
				sig = e.inner_text
				sig = sig.gsub(/\n/, "")
				sig = sig.gsub(/\s+/, " ")
				sig = sig.gsub(/\(\s+/, "(")
				sig = sig.gsub(/\s+\)/, ")")

				sig = sig.gsub(/\)\s+as\s+[^,]+$/, ")")

				# if there are optional params, break this thing up into multiple instances
				break_apart_optional_params(sig, is_js).each do | signature|
					desc = signature.gsub(/ as\s.*?(?=,|\)$)+/, "")
					desc = desc.gsub(/^[^(]+/, "")
					desc = desc.gsub(/,\s+/, ",")
					desc = nil if (desc == "()")

					# wrap the parameters with ${n:} where n is the parameter #
					if signature.match(/\$/)
						signature = signature.gsub(/\$/).with_index do |m, i|
							if is_js
								"${#{i+1}:"
							else
								"${#{i+1}:\\$"
							end
						end
						signature = signature.gsub(/,/, "},")
						signature = signature.gsub(/\)$/, "})")
					end

					function_name = $1 if signature =~ /([^(]+)\(/

					obj = {
						"content" => signature,
						"trigger" => function_name
					}

					obj['description'] = desc if (desc)

					final_functions[signature] = obj
				end
			end
		end
	end

	hydra.queue r
end

def build_functions(url, filename, is_js = false)
	functions = {}

	hydra = Typhoeus::Hydra.new
	response = Typhoeus::Request.post(url)
	html = Nokogiri::XML::Document.parse(response.body)
	html.remove_namespaces!
	html.xpath('//table[@class="api_table"]//tr/td/a/@href').each do |link|
		get_function(hydra, link.text, functions, is_js)
	end
	hydra.run

	# hydra = Typhoeus::Hydra.new
	# if (is_js)
	# 	get_function(hydra, "/8.0/xdmp.function-signature", functions, is_js)
	# else
	# 	get_function(hydra, "/8.0/xdmp:invoke-function", functions, is_js)
	# 	get_function(hydra, "/8.0/xdmp:set-response-code", functions, is_js)
	# 	get_function(hydra, "/8.0/xdmp:function-signature", functions, is_js)
	# end

	hydra.run

	pruned_functions = []
	functions.each_value do |v|
		pruned_functions.push(v)
	end

	print(pruned_functions)
	json = JSON.generate(pruned_functions.sort_by { |o| o['content'] })
	File.open(filename, "w") { |file| file.write(json) }
end

build_functions("#{HTTP_ROOT}/8.0/all", '../marklogic_builtins/ml-xquery-functions.json')
build_functions("#{HTTP_ROOT}/8.0/js/all", '../marklogic_builtins/ml-javascript-functions.json', true)